### ISSUE TITLE
Automatically optimize on special conditions 

### ISSUE THREAD

**Issue by adubovikov** [ClickHouse/ClickHouse#35836]
Clickhouse has special options "do_not_merge_across_partitions_select_final" for ReplaceMergeTree Engine, which allows make fast select on the "optimized" partitions. Unfortunate you have to run manually (crontab)  "OPTIMIZE TABLE ... PARTITION ... FINAL" . 

It will be great to have some options which will tell Clickhouse to compact older partitions on some conditions:

i.e.
```
CREATE TABLE select_final (t DateTime, x Int32, string String) ENGINE = ReplacingMergeTree() PARTITION BY toYYYYMMDD(t) ORDER BY (x, t)  SETTINGS optimize_partition_after = 3600
```
Clickhouse will check if a partition is older than 3600 seconds and has multiple parts and in such condition will do automatically optimization (compaction) 

P.S. we just discussed with @alexey-milovidov about this options :-)

**Comment by Slind14** [comment]
The logic we use for this prioritizes:

1. bytes of mergeable data (aggregating, collapsing tree)
2. max modified time for the partition

If we wouldn't need to manage this, it would be great.

**Comment by adubovikov** [comment]
I think we have bug here.

```
DROP TABLE select_final;
CREATE TABLE select_final (t DateTime, u Int32, x Int32, string String) ENGINE = ReplacingMergeTree(u) PARTITION BY toYYYYMMDD(t) ORDER BY (x, t) SETTINGS min_age_to_force_merge_seconds = 120;
INSERT INTO select_final SELECT toDate('2022-12-14'), toUnixTimestamp(NOW()), number, 'level 1' FROM numbers(100000000);
INSERT INTO select_final SELECT toDate('2022-12-14'), toUnixTimestamp(NOW()), number, 'level 2' FROM numbers(100000000);
INSERT INTO select_final SELECT toDate('2022-12-14'), toUnixTimestamp(NOW()), number, 'level 3' FROM numbers(100000000);

--- waiting for 120 seconds

select count(*) from system.parts where table = 'select_final' AND active=1\G;

---  result 7 parts - expecting 1

--- waiting another 120 seconds .... 

select count(*) from system.parts where table = 'select_final' AND active=1\G;

---  result 7 parts - expecting 1 

OPTIMIZE table select_final PARTITION 20221214 FINAL;

select count(*)  from system.parts where table = 'select_final' AND active=1\G;

---  result 1 part as expected.

DROP TABLE select_final;
````

```
SELECT VERSION()

Query id: 070564ea-ae6f-4cf1-8fdb-c5ef31720e32

┌─version()──┐
│ 22.11.2.30 │
└────────────┘

```

I think the settings min_age_to_force_merge_seconds doesn't work correct for ReplacingMergeTree.

**Comment by adubovikov** [comment]
@alexey-milovidov can you please confirm ?

**Comment by antonio2368** [comment]
I run the queries you provided and I could observe the correct behaviour for the setting.
Here are a couple of things to be aware of.
There are 2 "mods" of `min_age_to_force_merge_seconds` setting controlled by `min_age_to_force_merge_on_partition_only` (default set to `false`). 
If this is set to `true`, we will try to merge the entire partition if all parts in it are older than `min_age_to_force_merge_seconds`. Using this should show you a bigger drop in parts number and it's probably something you wanted to test.
If this is set to `false`, the merge will happen automatically only if the range of part CH picked is older than `min_age_to_force_merge_seconds`. You depend on the ranges that are picked, and the drops should be more gradual (older range is merged which creates new parts which will prevent some ranges to be merged until `min_age_to_force_merge_seconds` passes). This option is less greedy but it's good for cases when you still modify partition but you would like to merge some subset of older parts if possible.

Also, the age is referring to the age of parts. So if CH does merges for any other reason the `min_age_to_force_merge_seconds` will start counting from that point because those new parts will be newer than the initial ones. This, coupled with `min_age_to_force_merge_on_partition_only` not set to true is hard for you to verify the behavior using the exact timers.

I tried with your example, and I actually got the correct drops (not present when the settings are not used) in both cases so I'm not sure what went wrong in your case.
